src/java.base/share/classes/java/lang/Shutdown.java

```diff
@@ -157,37 +157,39 @@ static void halt(int status) {
      * which should pass a nonzero status code.
      */
     static void exit(int status) {
-        System.Logger log = getRuntimeExitLogger(); // Locate the logger without holding the lock;
+        logRuntimeExit(status);         // Log without holding the lock;
+
         synchronized (Shutdown.class) {
             /* Synchronize on the class object, causing any other thread
              * that attempts to initiate shutdown to stall indefinitely
              */
-            if (log != null) {
-                Throwable throwable = new Throwable("Runtime.exit(" + status + ")");
-                log.log(System.Logger.Level.DEBUG, "Runtime.exit() called with status: " + status,
-                        throwable);
-            }
             beforeHalt();
             runHooks();
             halt(status);
         }
     }
 
-    /* Locate and return the logger for Shutdown.exit, if it is functional and DEBUG enabled.
-     * Exceptions should not prevent System.exit; the exception is printed and otherwise ignored.
+    /* Locate the logger and log the Runtime.exit(status).
+     * Catch and ignore any and all exceptions.
      */
-    private static System.Logger getRuntimeExitLogger() {
+    private static void logRuntimeExit(int status) {
         try {
             System.Logger log = System.getLogger("java.lang.Runtime");
-            return (log.isLoggable(System.Logger.Level.DEBUG)) ? log : null;
+            if (log.isLoggable(System.Logger.Level.DEBUG)) {
+                Throwable throwable = new Throwable("Runtime.exit(" + status + ")");
+                log.log(System.Logger.Level.DEBUG, "Runtime.exit() called with status: " + status,
+                        throwable);
+            }
         } catch (Throwable throwable) {
-            // Exceptions from locating the Logger are printed but do not prevent exit
-            System.err.println("Runtime.exit() log finder failed with: " + throwable.getMessage());
+            try {
+                // Exceptions from the Logger are printed but do not prevent exit
+                System.err.println("Runtime.exit() logging failed: " + throwable.getMessage());
+            } catch (Throwable throwable2) {
+                // Ignore
+            }
```

Comment:
The update means that several threads racing to exit may result in more than one message being logged. I don't think that's a big issue but just wanted to make sure that this was deliberate.

Reply:
Yes, its racy, and intentional, but from a diagnostic view, the developer should know that multiple callers were trying to exit. All but one will hang.  
Consolidating the code in a single method is cleaner and with fewer interactions between separate code blocks.

Reply:
I can't tell if the thread identity will be included in the log the message?

I'm not so sure it is useful to allow competing exit calls to all log their intent. I would have kept the logging itself inside the locked region as per the original code.

Reply:
The main diagnostic is the logged message with the stack trace. The println version is present as a diagnostic if logging is mis-configured.

If the developer is looking to mitigate calls to Runtime.Exit, then they are advised to check all paths and will appreciate not having to wait for a different thread to win the race.

Reply:
But does that logging include the thread identity? If multiple threads can race to exit and all log, then the developer/user needs to know which logging came from which thread.

Reply:
> But does that logging include the thread identity? If multiple threads can race to exit and all log, then the developer/user needs to know which logging came from which thread.

That's really up to the Logger and its configuration. If j.u.logging is used then formatters can be configured to put the thread ID into the log records. With 3rd party logging libraries there seems to be several choices, like %t for the thread name.

The main usage for this logging is to be able to find code in tests, plugins, etc. that is calling System.exit and causing the test runner or container to exit. So I think it's less about "which thread" and more about "which code".

Reply:
System.Logger is a facade for arbitrary user-code, commonly the log4J bridge. I routinely observe log4J throwing SecurityException's from deep within its implementation (since it is only partially implemented to handle the security manager)